File: D12/d12.py
def f(p):
  x,y=0,0 # x - east, y - north
  v='ESWN'
  d='E'
  for c in p:
    n=int(c[1:])
    if   c[0]=='E': x+=n
    elif c[0]=='W': x-=n
    elif c[0]=='N': y+=n
    elif c[0]=='S': y-=n
    elif c[0]=='R':
      if   n==90:  d=v[(v.find(d)+1)%len(v)]
      elif n==180: d=v[(v.find(d)+2)%len(v)]
      elif n==270: d=v[(v.find(d)+3)%len(v)]
    elif c[0]=='L':
      if   n==90:  d=v[(v.find(d)+len(v)-1)%len(v)]
      elif n==180: d=v[(v.find(d)+len(v)-2)%len(v)]
      elif n==270: d=v[(v.find(d)+len(v)-3)%len(v)]
    elif c[0]=='F':
      if   d=='E': x+=n
      elif d=='W': x-=n
      elif d=='N': y+=n
      elif d=='S': y-=n
  return abs(x)+abs(y)

def g(p):
  x,y=0,0 # x - east, y - north
  wpx,wpy=10,1 # 'E','N'
  for c in p:
    n=int(c[1:])
    if   c[0]=='E': wpx+=n
    elif c[0]=='W': wpx-=n
    elif c[0]=='N': wpy+=n
    elif c[0]=='S': wpy-=n
    elif c[0]=='R':
      if   n==90:  wpx,wpy = wpy,-wpx
      elif n==180: wpx,wpy = -wpx,-wpy
      elif n==270: wpx,wpy = -wpy,wpx
      else:
        print(c)
    elif c[0]=='L':
      if   n==90:  wpx,wpy = -wpy,wpx
      elif n==180: wpx,wpy = -wpx,-wpy
      elif n==270: wpx,wpy = wpy,-wpx
    elif c[0]=='F':
      x+=n*wpx
      y+=n*wpy
  return abs(x)+abs(y)

p = open("d12.txt","rt").read().splitlines()
print(f(p))
print(g(p))

File: D12/d12.go
package D12

import (
	"os"
	"path"
	"bufio"
	"strconv"
	"fmt"
	"math"
)

const (
	EAST = iota
	NORTH
	WEST
	SOUTH
)

type Ship struct {
	X int64;
	Y int64;
	Facing int64;
}
type Instr struct {
	Op rune
	Arg int64
}


var instr []Instr

func Part1() {
	cwd, _ := os.Getwd()
	file, _ := os.Open(path.Join(cwd, "D12/d12.txt"))
	scanner := bufio.NewScanner(file)
	ship := Ship{}
	for scanner.Scan() {
		val := scanner.Text()
		op := rune(val[0])
		_arg, _ := strconv.Atoi(val[1:])
		arg := int64(_arg)
		switch op {
		case 'N':
			ship.Y += arg
		case 'W':
			ship.X -= arg
		case 'S':
			ship.Y -= arg
		case 'E':
			ship.X += arg
		case 'F':
			switch ship.Facing {
			case EAST:
				ship.X += arg
			case WEST:
				ship.X -= arg
			case SOUTH:
				ship.Y -= arg
			case NORTH:
				ship.Y += arg
			}
		case 'L':
			ship.Facing = (ship.Facing + arg / 90) % 4
		case 'R':
			ship.Facing = (ship.Facing + (360-arg)/ 90) % 4
		}
		instr = append(instr, Instr{Op: op, Arg: arg})
	}
	if ship.X < 0 {
		ship.X = -ship.X
	}
	if ship.Y < 0 {
		ship.Y = -ship.Y
	}
	fmt.Println(ship.X + ship.Y)
}

func Part2() {
	wp := Ship{X: 10, Y: 1}
	ship := Ship{}
	for _, inst := range instr {
		op, arg := inst.Op, inst.Arg
		rad := math.Pi * (float64(arg) / 180)
		switch op {
		case 'N':
			wp.Y += arg
		case 'W':
			wp.X -= arg
		case 'S':
			wp.Y -= arg
		case 'E':
			wp.X += arg
		case 'F': // move the ship toward the waypoint
			ship.X = ship.X + wp.X * arg
			ship.Y = ship.Y + wp.Y * arg
		case 'L':
			leng := math.Sqrt(float64(wp.X * wp.X + wp.Y * wp.Y))
			wp.X = int64(float64(wp.X) * math.Cos(rad) - float64(wp.Y) * math.Sin(rad))
			wp.Y = int64(float64(wp.X) * math.Sin(rad) + float64(wp.Y) * math.Cos(rad))
			ratio := leng / math.Sqrt(float64(wp.X*wp.X + wp.Y *wp.Y))
			wp.X *= int64(ratio)
			wp.Y *= int64(ratio)
		case 'R':
			leng := math.Sqrt(float64(wp.X * wp.X + wp.Y * wp.Y))
			wp.X = int64(float64(wp.X) * math.Cos(-rad) - float64(wp.Y) * math.Sin(-rad))
			wp.Y = int64(float64(wp.X) * math.Sin(-rad) + float64(wp.Y) * math.Cos(-rad))
			ratio := leng / math.Sqrt(float64(wp.X*wp.X + wp.Y *wp.Y))
			wp.X *= int64(ratio)
			wp.Y *= int64(ratio)
		}
		fmt.Println(op, arg, ship)
		fmt.Println(wp)
		fmt.Println()
	}
	if ship.X < 0 {
		ship.X = -ship.X
	}
	if ship.Y < 0 {
		ship.Y = -ship.Y
	}
	// fmt.Println(ship.X, ship.Y)
	fmt.Println(ship.X + ship.Y)
}